flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/DeltaJoinTest.scala

```diff
@@ -0,0 +1,574 @@
+        "on src1.a2 = src2.b2")
+  }
+
+  @Test
+  def testWithNonEquiCondition1(): Unit = {
+    util.verifyRelPlanInsert(
+      "insert into snk select * from src1 join src2 " +
+        "on src1.a1 = src2.b1 " +
+        "and src1.a2 = src2.b2 " +
+        "and src2.b0 > src1.a0")
+  }
+
+  @Test
+  def testWithNonEquiCondition2(): Unit = {
+    // could not optimize into delta join because there is a calc between join and source
+    util.verifyRelPlanInsert(
+      "insert into snk select * from src1 join src2 " +
+        "on src1.a1 = src2.b1 " +
+        "and src1.a2 = src2.b2 " +
+        "and src2.b0 > src1.a0 " +
+        "and src2.b2 <> 'Hello' " +
+        "and src1.a0 > 99")
+  }
+
+  @Test
+  def testJsonPlanWithTableHints(): Unit = {
+    util.verifyJsonPlan(
+      "insert into snk select * from src1 /*+ OPTIONS('failing-source'='true') */" +
+        "join src2 /*+ OPTIONS('failing-source'='true') */" +
+        "on src1.a1 = src2.b1 " +
+        "and src1.a2 = src2.b2 " +
+        "and src2.b0 > src1.a0")
+  }
+
+  @Test
+  def testProjectFieldsBeforeJoin(): Unit = {
+    // could not optimize into delta join because the source has ProjectPushDownSpec
+    util.verifyRelPlanInsert(
+      "insert into snk(l0, l1, l2, r0, r2, r1) " +
+        "select * from ( " +
+        "  select a0, a1, a2 from src1" +
+        ") tmp join src2 " +
+        "on tmp.a1 = src2.b1 " +
+        "and tmp.a2 = src2.b2")
+  }
+
+  @Test
+  def testProjectFieldsAfterJoin(): Unit = {
+    util.verifyRelPlanInsert(
+      "insert into snk select a0, a1 + 1.1, a2, a3, b0, b2, b1 from src1 join src2 " +
+        "on src1.a1 = src2.b1 " +
+        "and src1.a2 = src2.b2")
+  }
+
+  @Test
+  def testFilterFieldsBeforeJoin(): Unit = {
+    // could not optimize into delta join because there is a calc between source and join
+    util.verifyRelPlanInsert(
+      "insert into snk select * from (  " +
+        "  select * from src1 where a1 > 1.1 " +
+        ") tmp join src2 " +
+        "on tmp.a1 = src2.b1 " +
+        "and tmp.a2 = src2.b2")
+  }
+
+  @Test
+  def testFilterFieldsAfterJoin(): Unit = {
+    util.verifyRelPlanInsert(
+      "insert into snk select * from src1 join src2 " +
+        "on src1.a1 = src2.b1 " +
+        "and src1.a2 = src2.b2 " +
+        "where a3 > b0")
+  }
+
+  @Test
+  def testMultiRootsWithReusingJoinView(): Unit = {
+    util.tableConfig.set(
+      OptimizerConfigOptions.TABLE_OPTIMIZER_REUSE_OPTIMIZE_BLOCK_WITH_DIGEST_ENABLED,
+      Boolean.box(true))
+
+    util.tableEnv.executeSql("create table snk2 like snk")
+
+    util.tableEnv.executeSql(
+      "create temporary view mv as " +
+        "select * from src1 join src2 " +
+        "on src1.a1 = src2.b1 " +
+        "and src1.a2 = src2.b2")
+
+    val stmt = tEnv.createStatementSet()
+    stmt.addInsertSql("insert into snk select * from mv")
+    stmt.addInsertSql("insert into snk2 select * from mv")
+
+    util.verifyRelPlan(stmt)
```

Comment:
Replace this with `verifyExecPlan` while removing the TODO.
The others are same.